### PR TITLE
fix: Correct v2 listing of more than 1000 objects

### LINKED ISSUE #578: listObjectsV2 causes endless loop > 1000 items
Using the new (default) listObjectsV2 to list objects causes an endless loop when there are more than 1000 items returned by the server and _recursive_ is set to `true`. The iterator will list the first 1000 items over and over again.

I guess there is some error in the continuation token logic in the java client.

To reproduce, create 1001 items in a bucket and use the simplistic code below to list them:
```java
    private static void testEndlessLoop(){

        MinioClient minioClient = null;
        try {
            minioClient = new MinioClient(
                    "http://myminio.lan:9000",
                    accessKey,
                    accessSecret);


            Iterable<Result<Item>> iterable = minioClient.listObjects(
                    "mybucket",
                    "some/prefix/",
                    true,
                    false);  // if set to true it works, using V1
            

            iterable.forEach(o -> {
                try {
                    System.out.println(o.get().objectName());
                } catch (Exception e) {
                    e.printStackTrace();
                }
            });
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
```

If you have trouble reproducing it on your end let me know.

**Versions**
`minio-java 3.0.4`
`minio  RELEASE.2017-05-05T01-14-51Z `

### PR DESCRIPTION
**Pull request by vadmeste**
Fix listing objects (infinite loop) when there is more than
1000 objects in the server.

Fixes #578

### CODE REVIEW

File: functional/FunctionalTest.java [hunk 1]
```diff
@@ -668,10 +668,41 @@ public static void listObject_test4() throws Exception {
   }
 
   /**
-   * Test: listObjects(bucketName, final String prefix, final boolean recursive, final boolean useVersion1).
+   * Test: listObjects(bucketName, final String prefix, final boolean recursive).
    */
   public static void listObject_test5() throws Exception {
     int i;
+    int totalObjsNr = 1050;
```
Comment (balamurugana) [inline functional/FunctionalTest.java]: `objCount` or `totalObjects` is better

File: functional/FunctionalTest.java [hunk 2]
```diff
@@ -668,10 +668,41 @@ public static void listObject_test4() throws Exception {
...
+
+    System.out.println("Test: listObjects(final String bucketName, final String prefix, final boolean recursive)");
+    String[] filenames = new String[totalObjsNr];
+    for (i = 0; i < totalObjsNr; i++) {
+      String filename = createFile(1);
+      client.putObject(bucketName, filename, filename);
+      Files.delete(Paths.get(filename));
+      filenames[i] = filename;
```
Comment (balamurugana) [inline functional/FunctionalTest.java]: you could do below optimization
```java
String baseFilename = createFile(1);
for (...) {
  objectNames[i] = baseFilename + "-" + i;
  client.putObject(bucketName, objectNames[i], filename);
}
Files.delete(Path.get(baseFilename));
```

File: functional/FunctionalTest.java [hunk 3]
```diff
@@ -668,10 +668,41 @@ public static void listObject_test4() throws Exception {
   }
 
   /**
-   * Test: listObjects(bucketName, final String prefix, final boolean recursive, final boolean useVersion1).
+   * Test: listObjects(bucketName, final String prefix, final boolean recursive).
```
Comment (balamurugana) [inline functional/FunctionalTest.java]: You could indicate the test behavior by prefixing `Test: recursive: `

File: functional/FunctionalTest.java [hunk 4]
```diff
@@ -668,10 +668,41 @@ public static void listObject_test4() throws Exception {
...
    */
   public static void listObject_test5() throws Exception {
     int i;
+    int totalObjsNr = 1050;
+
+    System.out.println("Test: listObjects(final String bucketName, final String prefix, final boolean recursive)");
```
Comment (balamurugana) [inline functional/FunctionalTest.java]: You could indicate the test behavior by prefixing `Test: recursive: `

File: functional/FunctionalTest.java [hunk 5]
```diff
@@ -668,10 +668,41 @@ public static void listObject_test4() throws Exception {
...
+    }
+
+    i = 0;
+    for (Result<?> r : client.listObjects(bucketName, "minio", true)) {
+      ignore(i++, r.get());
+    }
+
+    // Check the number of uploaded objects
+    if (i != totalObjsNr) {
+      throw new Exception("[FAILED] Test: listObject_test5(), number of items found: " + i);
```
Comment (balamurugana) [inline functional/FunctionalTest.java]: You could indicate `expected` in the error message `"[FAILED] Test: listObject_test5(), expected object count:  " + objCount + ", got: " + i`

**Comment by vadmeste** [comment]
All your comments are addressed @balamurugana, except the commit message, your suggestion was long (>50 chars) so I tried a shorter one.

**Comment by harshavardhana** [comment]
@vadmeste 

```
https://repo1.maven.org/maven2/org/abego/treelayout/org.abego.treelayout.core/1.0.1/org.abego.treelayout.core-1.0.1.jar
[ant:checkstyle] C:\projects\minio-java\functional\FunctionalTest.java:677: Line is longer than 120 characters (found 126).
[ant:checkstyle] C:\projects\minio-java\functional\FunctionalTest.java:694: Line is longer than 120 characters (found 132).
:api:checkstyleMain FAILED
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':api:checkstyleMain'.
> Checkstyle rule violations were found. See the report at: file:///C:/projects/minio-java/api/build/reports/checkstyle/main.html
* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
BUILD FAILED
Total time: 55.236 secs
Command
``` 

Some stylecheck failures.

**Comment by vadmeste** [comment]
Fix style done.

**Comment by vadmeste** [comment]
> I would suggest you to prefix fix: in the commit message. Thats the convention followed.

Fixed

**Pull request merged**